### PR TITLE
push to stable

### PR DESCRIPTION
**Pull request by vanhauser-thc**
here we go :)
this is missing a fix for qemu static from @andreafioraldi and afl-cmin -f from @hexcoder- ... anything else?

### CODE REVIEW

File: include/config.h [hunk 1]
```diff
@@ -36,6 +34,38 @@
...
 
+/* CMPLOG/REDQUEEN TUNING
+ *
+ * Here you can modify tuning and solving options for CMPLOG.
+ * Note that these are run-time options for afl-fuzz, no target
+ * recompilation required.
+ *
+ */
+
+/* Enable arithmetic compare solving for both path */
```
Comment (hexcoder-) [inline include/config.h:45]: question:
for both path
->
for both branches (true and false)
?

File: include/config.h [hunk 2]
```diff
@@ -36,6 +34,38 @@
...
+#define CMPLOG_SOLVE_ARITHMETIC
+
+/* Enable transform following (XOR/ADD/SUB manipulations, hex en/decoding) */
+#define CMPLOG_SOLVE_TRANSFORM
+
+/* if TRANSFORM is enabled, this additionally enables base64 en/decoding */
+// #define CMPLOG_SOLVE_TRANSFORM_BASE64
+
+/* If a redqueen pass finds more than one solve, try to combine them? */
```
Comment (hexcoder-) [inline include/config.h:54]: solve
->
solution
?

File: include/envs.h [hunk 3]
```diff
@@ -42,11 +43,13 @@ static char *afl_environment_variables[] = {
     "AFL_DEBUG_GDB",
     "AFL_DISABLE_TRIM",
     "AFL_DONT_OPTIMIZE",
+    "AFL_DRIVER_STDERR_DUPLICATE_FILENAME",
```
Comment (hexcoder-) [inline include/envs.h:47]: undocumented in docs/env_variables.md

File: include/envs.h [hunk 4]
```diff
@@ -42,11 +43,13 @@ static char *afl_environment_variables[] = {
...
     "AFL_DUMB_FORKSRV",
     "AFL_ENTRYPOINT",
     "AFL_EXIT_WHEN_DONE",
     "AFL_FAST_CAL",
     "AFL_FORCE_UI",
+    "AFL_FUZZER_ARGS",  // oss-fuzz
```
Comment (hexcoder-) [inline include/envs.h:53]: dito, undocumented.

File: include/envs.h [hunk 5]
```diff
@@ -162,6 +165,7 @@ static char *afl_environment_variables[] = {
     "AFL_WINE_PATH",
     "AFL_NO_SNAPSHOT",
     "AFL_EXPAND_HAVOC_NOW",
+    "AFL_USE_QASAN",
     NULL
```
Comment (hexcoder-) [inline include/envs.h:171]: `AFL_QEMU_FORCE_DFL` is missing?

File: instrumentation/afl-compiler-rt.o.c [hunk 6]
```diff
@@ -1090,7 +1090,7 @@ __attribute__((constructor(0))) void __afl_auto_first(void) {
   if (getenv("AFL_DISABLE_LLVM_INSTRUMENTATION")) return;
   u8 *ptr;
 
-  ptr = (u8 *)malloc(1024000);
+  ptr = (u8 *)malloc(2097152);
```
Comment (hexcoder-) [inline instrumentation/afl-compiler-rt.o.c:1093]: suggest using `MAP_INITIAL_SIZE` instead of 2097152 here.

**Comment by andreafioraldi** [comment]
pushed the static build fix to stable by mistake, will sync the commit in dev after merge

File: qemu_mode/libqasan/README.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# QEMU AddressSanitizer Runtime
+
+This library is the injected runtime used by QEMU AddressSanitizer (QASan).
+
+The original repository is [here](https://github.com/andreafioraldi/qasan).
+
+The version embedded in qemuafl is an updated version of just the usermode part and this runtime in injected via LD_PRELOAD (so works just for dynamically linked binaries).
```
Comment (hexcoder-) [inline qemu_mode/libqasan/README.md]: in injected
->
is injected
?

File: README.md [hunk 8]
```diff
@@ -55,7 +55,7 @@ behaviours and defaults:
     * a caching of testcases can now be performed and can be modified by
       editing config.h for TESTCASE_CACHE or by specifying the env variable
       `AFL_TESTCACHE_SIZE` (in MB). Good values are between 50-500 (default: 50).
-    * -M mains do not perform trimming
+    * -M mains does not perform trimming
```
Comment (vanhauser-thc) [inline README.md]: mains is plural so "do" was correct

File: README.md [hunk 9]
```diff
@@ -717,11 +722,53 @@ Crash processing
...
+
+3. Also randomize the afl-fuzz runtime options, e.g.
+  * 60% for `AFL_DISABLE_TRIM`
+  * 50% use a dictionary generated by `AFL_LLVM_DICT2FILE`
+  * 50% use MOpt (`-L 0`)
+  * 40% for `AFL_EXPAND_HAVOC_NOW`
+  * 30% for old queue processing (`-Z`)
+  * for CMPLOG targets, 60% for `-l 2`, 40% for `-l 3`
+
+4. Do *not* run any `-M` modes, just running `-S` modes is better for CI fuzzing.
```
Comment (domenukk) [inline README.md:753]: Can we add an explanation here, why?

File: README.md [hunk 10]
```diff
@@ -732,14 +779,18 @@ feature by doing:
...
+check out the snapshot feature! :-)
 If possible you should use the persistent mode, see [qemu_mode/README.persistent.md](qemu_mode/README.persistent.md).
 The mode is approximately 2-5x slower than compile-time instrumentation, and is
 less conducive to parallelization.
 
 If [afl-dyninst](https://github.com/vanhauser-thc/afl-dyninst) works for
 your binary, then you can use afl-fuzz normally and it will have twice
 the speed compared to qemu_mode (but slower than persistent mode).
+Note that several other binary rewriters exist, all with their advantages and
+caveats. As rewriting a binary is much faster than Qemu this is a highly
```
Comment (domenukk) [inline README.md]: I would usually recommend qemu mode - it's maintained, not super slow, but has cmplog, cmpcov support (?)
Maybe not "highly recommended" but "may be a better approach for your use case, due to higher throughput"

File: docs/env_variables.md [hunk 11]
```diff
@@ -431,13 +436,18 @@ checks or alter some of the more exotic semantics of the tool:
...
+    The host and port are configurable with `AFL_STATSD_HOST` and `AFL_STATSD_PORT` respectively.
+    To enable tags (banner and afl_version) you should provide `AFL_STATSD_TAGS_FLAVOR` that matches
+    your StatsD server (see `AFL_STATSD_TAGS_FLAVOR`)
+
+  - Setting `AFL_STATSD_TAGS_FLAVOR` to one of `dogstatsd`, `librato`, `signalfx` or `influxdb`
+    allows you to add tags to your fuzzing instances. This is especially useful when running
+    multiple instances (`-M/-S` for example). Applied tags are `banner` and `afl_version`.
+    `banner` corresponds to the name of the fuzzer provided through `-M/-S`.
+    `afl_version` corresponds to the currently running afl version (e.g `++3.0c`).
+    Default (empty/non present) will add no tags to the metrics.
```
Comment (domenukk) [inline docs/env_variables.md:453]: Should this link to the new doc about statsd? (`docs/rpc_statsd.md`)

File: qemu_mode/libqasan/README.md [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+The original repository is [here](https://github.com/andreafioraldi/qasan).
+
+The version embedded in qemuafl is an updated version of just the usermode part and this runtime is injected via LD_PRELOAD (so works just for dynamically linked binaries).
+
+The usage is super simple, just set the env var `AFL_USE_QASAN=1` when fuzzing in qemu mode (-Q). afl-fuzz will automatically set AFL_PRELOAD to load this library and enable the QASan instrumentation in afl-qemu-trace.
+
+For debugging purposes, we still suggest to run the original QASan as the stacktrace support for ARM (just a debug feature, it does not affect the bug finding capabilities during fuzzing) is WIP.
+
+### When I should use QASan?
```
Comment (domenukk) [inline qemu_mode/libqasan/README.md]: Italians switch `should I` and `I should` :P

**Review by andreafioraldi — APPROVED** [review]
lgtm

**Review by domenukk — APPROVED** [review]

File: qemu_mode/libqasan/dlmalloc.c [hunk 13]
```diff
@@ -3916,8 +3916,11 @@ static void internal_malloc_stats(mstate m) {
           clear_smallmap(M, I);                                 \
                                                                 \
         } else if (RTCHECK(B == smallbin_at(M, I) ||            \
+<<<<<<< HEAD
+=======
                                                                 \
```
Comment (vanhauser-thc) [inline qemu_mode/libqasan/dlmalloc.c:3919]: ah ... this looks like it went very wrong ...

**Comment by vanhauser-thc** [comment]
@hexcoder- done

**Pull request merged**